File: 2019/day19/cpp/src/beam.cpp
#include "beam.hpp"
#include <vector>

Beam::Beam(std::string file_name) : m_file_name(file_name), m_intcode() {
  m_intcode = Intcode(file_name, {}, true, false);
}

int Beam::part_one(int grid_size) {

  int nAsteroids(0);
  for (int x = 0; x < grid_size; x++) {
    for (int y = 0; y < grid_size; y++) {
      auto output = m_intcode({x,y});
      nAsteroids += std::get<1>(output);
    }
  }
  return nAsteroids;
}

int Beam::part_two(int square_size) {

  // Bottom left  corner
  std::vector<int64_t> bot_left({0,square_size-1});

  while (true) {
    // If bottom left corner has a detection
    if (std::get<1>(m_intcode(bot_left))) {
      // Check if top right has a detection - done
      auto top_right = std::vector<int64_t>({bot_left[0]+(square_size-1),
                                             bot_left[1] -(square_size-1)});
      if (std::get<1>(m_intcode(top_right))) {
        return bot_left[0]*10000 + top_right[1];
      // Otherwise move down
      } else {
        bot_left[1]++;
      }
    // Otherwise move to the right
    } else {
      bot_left[0]++;
    }
  }
}

File: 2019/day19/beam.py
from os.path import dirname
from os.path import realpath
from os.path import join
import sys

sys.path.append(join(dirname(realpath(__file__)), *[".."]))
from day5.intcode import Intcode


class Beam:
    """Tractor beam class"""

    def __init__(self, fileName):
        """Initialize Beam with the intcode computer"""

        self.fileName = fileName
        self.ic = Intcode(fileName, verbose=False, reset=True)

    def part_one(self, grid_size=50):
        """Loop (x,y) in a N x N grid and sum up the output"""
        return sum(
            [
                self.ic([x, y])[1]
                for x in range(grid_size)
                for y in range(grid_size)
            ]
        )

    def part_two(self, square_size=100):
        """Move a square box of size NxN down and right until bottom
        left and top right corners both return output = 1"""
        # Bottom left X,Y coordinates
        bot_left = [0, square_size - 1]

        # Keep searching until we find the square box
        while True:

            # Check bottom left corner returns output 1
            if self.ic(bot_left)[1]:
                top_right = [
                    bot_left[0] + (square_size - 1),
                    bot_left[1] - (square_size - 1),
                ]
                # Check if also top right corner returns output 1
                if self.ic(top_right)[1]:
                    return bot_left[0] * 10000 + top_right[1]
                # Otherwise move down
                else:
                    bot_left[1] += 1

            # Otherwise move to the right
            else:
                bot_left[0] += 1
